mlfinlab/portfolio_optimization/hcaa.py

```diff
@@ -0,0 +1,421 @@
+        return cluster_expected_shortfall
+
+    def _get_cluster_conditional_drawdown_at_risk(self, asset_returns, covariance, confidence_level, cluster_indices):
+        '''
+        Calculate cluster conditional drawdown at risk.
+
+        :param asset_returns: (pd.DataFrame) historical asset returns
+        :param covariance: (pd.DataFrame) covariance matrix of assets
+        :param confidence_level: (float) the confidence level (alpha)
+        :param cluster_indices: (list) list of asset indices for the cluster
+        :return: (float) CDD of the cluster
+        '''
+
+        cluster_asset_returns = asset_returns.iloc[:, cluster_indices]
+        cluster_covariance = covariance.iloc[cluster_indices, cluster_indices]
+        parity_w = self._get_inverse_variance_weights(cluster_covariance)
+        portfolio_returns = cluster_asset_returns @ parity_w
+        cluster_conditional_drawdown = self.risk_metrics.calculate_conditional_drawdown_risk(returns=portfolio_returns,
+                                                                                             confidence_level=confidence_level)
+        return cluster_conditional_drawdown
+
+    def _recursive_bisection(self,
+                             expected_asset_returns,
+                             asset_returns,
+                             covariance_matrix,
+                             assets,
+                             allocation_metric,
+                             confidence_level):
+        # pylint: disable=bad-continuation, too-many-locals
+        '''
+        Recursively assign weights to the clusters - ultimately assigning weights to the individual assets.
+
+        :param expected_asset_returns: (list) a list of mean asset returns (mu)
+        :param asset_returns: (pd.DataFrame) historical asset returns
+        :param covariance_matrix: (pd.DataFrame) the covariance matrix
+        :param assets: (list) list of asset names in the portfolio
+        :param allocation_metric: (str) the metric used for calculating weight allocations
+        :param confidence_level: (float) the confidence level (alpha)
+        '''
+
+        self.weights = pd.Series(1, index=self.ordered_indices)
+        clustered_alphas = [self.ordered_indices]
+
+        while clustered_alphas:
+            clustered_alphas = [cluster[start:end]
+                                for cluster in clustered_alphas
+                                for start, end in ((0, len(cluster) // 2), (len(cluster) // 2, len(cluster)))
+                                if len(cluster) > 1]
+
+            for subcluster in range(0, len(clustered_alphas), 2):
+                left_cluster = clustered_alphas[subcluster]
+                right_cluster = clustered_alphas[subcluster + 1]
+
+                # Calculate allocation factor based on the metric
+                if allocation_metric == 'minimum_variance':
```

Comment:
I see here that these allocation_metrics could be coded up as a Class object. I think this will improve the readability and the extensibility of the code. Would it be worth it to invest the time developing them?

Reply:
@Jackal08 I already created the RiskMetrics class for these metrics

Reply:
Will explore this topic over slack.